bids/modeling/statsmodels.py

```diff
@@ -0,0 +1,627 @@
+        self.name = name
+        self.model = model or {}
+        self.transformations = transformations or []
+        self.contrasts = contrasts or []
+        self.dummy_contrasts = dummy_contrasts
+        self.groupby = groupby or []
+        self._collections = []
+        self._group_data = []
+        self.children = []
+        self.parents = []
+
+    @staticmethod
+    def _build_groups(objects, groupby):
+        """Group list of objects into bins defined by specified entities.
+
+        Parameters
+        ----------
+        objects : list
+            List of objects containing an .entities dictionary as an attribute.
+        groupby : list of str
+            List of strings indicating which entities to group on.
+
+        Returns
+        -------
+        A dictionary, where the keys are tuples of tuples and the values are lists
+        containing subsets of the original object list. Each outer tuple of the keys
+        represents a particular entity/value combination; each inner tuple has two
+        elements, where the first is the name of the entity and the second is the value.
+
+        Notes
+        -----
+        * The passed objects can be of any type, but alll elements in the list must
+        have an implemented .entities dictionary.
+        * Objects that do not have defined values for all of the requested
+        `entities` will be silently ignored.
+
+        Examples
+        --------
+        >>> bidsfile = namedtuple('bidsfile', ('entities',))
+        >>> bidsfile1 = bidsfile(entities={'subject': '01'})
+        >>> bidsfile2 = bidsfile(entities={'subject': '02'})
+        >>> groups = {(('subject', '01'),): [bidsfile1],
+        ...           (('subject', '02'),): [bidsfile2]}
+        >>> BIDSStatsModelsNode._build_groups(
+        ...     [bidsfile1, bidsfile2],
+        ...     ['subject']) == groups
+        True
+        """
+        if not groupby:
+            return {(): objects}
+
+        groups = defaultdict(list)
+
+        # sanitize grouping entities, otherwise weird things can happen
+        groupby = list(set(groupby) & VALID_GROUPING_ENTITIES)
+
+        # Get unique values in each grouping variable
+        entities = [obj.entities for obj in objects]
+        df = pd.DataFrame.from_records(entities).loc[:, groupby]
+        unique_vals = {col: df[col].dropna().unique().tolist() for col in groupby}
+
+        # Note: we can't just naively bucket objects based on the values of the
+        # grouping entities, because an object may have undefined values for
+        # one or more grouping variables. So we "fill in" the missing values by
+        # Looping over objects. For each one, identify all grouping variables
+        # with missing values. Loop over the cartesian product of the unique
+        # values of the missing entities, combining the permutation values with
+        # the base (present) grouping entities. Sort on this to produce the
+        # group key.
```

Comment:
The case that prompted doing it this way (I was initially doing it the simpler way) is unfortunately almost universal: `inputs` passed to `BIDSStatsModelsNode.run()` will always have a `"contrast"` entity defined (because that's added when the `ContrastInfo` objects are created for output), but collections loaded from the BIDS dataset will never have them (because they're meaningless outside of the modeling context).

The other non-crazy use case where this will happen is when you're bridging levels—e.g., if you pass outputs from the run level to a node defined at dataset level, then variables extracted from `participants.tsv` won't have defined values for `run`, `session`, etc.